theano/gof/destroyhandler.py

```diff
@@ -1016,6 +1016,8 @@ def orderings(self, fgraph):
                     # add the rule: app must be preceded by all other Apply instances that
                     # depend on destroyed_input
                     root_clients = OrderedSet()
+                    if app in rval:
+                        root_clients.update(rval[app])
```

Comment:
I didn't had the time to do the full review of the destroyhandler mechanism, but a local review of this diff gave us (me and @lamblin) that this PR is good to merge.

One small comment that would do a small code clean up and remove some potential slowdown when the change isn't needed:

Can you remove the 2 new lines and move line 1018 after line 952 (so between the 2 for loop). This will remove the need for the extra if, so no slow down will happen.

At the same time, can you remove one indentation level of ilne 1025 and 1026?

After that small change, we can merge the PR. If you can do it quickly, we want to make an release candidate of Theano and would like to include this fix.

thanks for all the work on this bugfix!